realsense2_camera/src/base_realsense_node.cpp

```diff
@@ -828,7 +839,47 @@ bool BaseRealSenseNode::shouldPublishCameraInfo(const stream_index_pair& sip)
     return (stream != RS2_STREAM_SAFETY && stream != RS2_STREAM_OCCUPANCY && stream != RS2_STREAM_LABELED_POINT_CLOUD);
 }
 
-void BaseRealSenseNode::publishLabeledPointCloud(rs2::labeled_points lpc, const rclcpp::Time& t){
+void BaseRealSenseNode::publishOccupancyFrame(rs2::frame f, const rclcpp::Time& t)
+{
+    if(!_occupancy_publisher || 0 == _occupancy_publisher->get_subscription_count())
+        return;
+
+    ROS_DEBUG("Publishing Occupancy GridCells Frame");
+
+    // get frame bytes and frame metadata relevant info
+    auto frame_as_uint8_arr = (uint8_t*)f.get_data();
+    auto cols = static_cast<int>(f.get_frame_metadata(RS2_FRAME_METADATA_OCCUPANCY_GRID_COLUMNS));
+    auto rows = static_cast<int>(f.get_frame_metadata(RS2_FRAME_METADATA_OCCUPANCY_GRID_ROWS));
+    auto cell_size = static_cast<float>(f.get_frame_metadata(RS2_FRAME_METADATA_OCCUPANCY_CELL_SIZE) / 100.0f); // convert to meters
+
+    // create GridCells msg and start filling it
+    nav_msgs::msg::GridCells msg;
+    msg.header.stamp = t;
+    msg.header.frame_id = FRAME_ID(OCCUPANCY);
+    msg.cell_width = cell_size;
+    msg.cell_height = cell_size;
+
+    for (auto i = 0; i < cols * rows; ++i)
+    {
+        // check if current cell is occupied
+        // each byte include 8 bits <--> 8 cells
+        if ((frame_as_uint8_arr[i / 8U] & (1U << i % 8)) != 0) 
+        {
+            // find x,y,z positions of current index
+            geometry_msgs::msg::Point p3d;
+            uint32_t row = (i / cols);
+            uint32_t col = (i % cols);
+            p3d.x = (cell_size * static_cast<float>(rows)) - cell_size * (static_cast<float>(row) + 0.5f);
```

Comment:
I would expect x axis will use columns and not rows, can you explain this line?

Reply:
in ROS CS, X is forward, Y is left, and Z is up.
in AICV algo ROWS represents the height is and COLS represents the width.
![image](https://github.com/IntelRealSense/realsense-ros/assets/99127997/c9f584c6-4708-48e8-be1d-6cafe9dcdc1f)